### PR TITLE
Upgrade to PHP 8

### PR DESCRIPTION
**Pull request by David-Denny**

### CODE REVIEW

File: src/Model/ModelTrait.php [hunk 1]
```diff
@@ -60,19 +63,20 @@ protected function loadData(array $data, $update = false, $replaceCollections =
...
-                        $isCollection &&
-                        $update &&
-                        (
-                            empty($replaceCollections) ||               // don't add if we're replacing all collections
-                            empty($replaceCollections[$this->mbLcfirst($prop)])  // don't add if this collection has been marked for replacing
+                        $isCollection
+                        && $update
+                        && (
+                            empty($replaceCollections)              // don't add if we're replacing all collections
+                            || empty($data[$this->mbLcfirst($prop)] // don't add if this collection has been marked for replacing
```
Comment (downsider) [inline src/Model/ModelTrait.php]: $replaceCollections is meant to be an array, so we can revert this to it's previous form (including moving the operators back to the end of the line)

File: src/Model/ModelTrait.php [hunk 2]
```diff
@@ -37,8 +39,9 @@ trait ModelTrait
      * @param array $data
      * @param bool $update
      * @param bool $replaceCollections
+     * @throws \ReflectionException|ModelException
      */
-    protected function loadData(array $data, $update = false, $replaceCollections = false)
+    protected function loadData(array $data, bool $update = false, bool $replaceCollections = false)
```
Comment (downsider) [inline src/Model/ModelTrait.php]: `$replaceCollections` should be an array. Can you update the typehint, docblock and set the default to `[]` please?

File: src/Model/ModelTrait.php [hunk 3]
```diff
@@ -91,15 +95,15 @@ protected function loadData(array $data, $update = false, $replaceCollections =
...
         }
     }
 
     /**
      * @param array $data
      * @param bool $replaceCollections
+     * @throws \ReflectionException|ModelException
      */
-    public function updateData(array $data, $replaceCollections = false)
+    public function updateData(array $data, bool $replaceCollections = false)
```
Comment (downsider) [inline src/Model/ModelTrait.php]: same here

File: src/Model/ModelTrait.php [hunk 4]
```diff
@@ -140,13 +145,16 @@ private function getFirstParameter($method)
...
+     * @return mixed
+     * @throws ModelException
+     * @throws \ReflectionException
      */
-    private function constructClasses(\ReflectionParameter $param, $value, $isCollection = false)
+    private function constructClasses(\ReflectionParameter $param, mixed $value, bool $isCollection = false): mixed
     {
-        $class = $param->getClass();
-        if (!empty($class)) {
+        $class = $param->getType() && !$param->getType()->isBuiltin()
```
Comment (downsider) [inline src/Model/ModelTrait.php]: This gets more complex now that union types are a thing. For now, I think we need to check that `$param->getType()` returns a `ReflectionNamedType` object before checking `isBuiltIn()`

Might be worth setting `param->getType()` to a variable since we use it a lot

File: src/Model/ModelTrait.php [hunk 5]
```diff
@@ -140,13 +145,16 @@ private function getFirstParameter($method)
...
+            ? new \ReflectionClass($param->getType()->getName()) : null;
 
+        if (!empty($class)) {
```
Comment (downsider) [inline src/Model/ModelTrait.php]: Think I'd prefer to check if it's an instance of `ReflectionClass` rather than empty, so we're explicit about what we expect here

File: src/Model/ModelTrait.php (hunk 2 above)
Comment (David-Denny) [reply src/Model/ModelTrait.php]: Ah I understand where my confusion was, I thought $replaceCollections was a bool because the default was false. Fixed now

File: src/Model/ModelTrait.php (hunk 1 above)
Comment (David-Denny) [reply src/Model/ModelTrait.php]: Fixed

File: src/Model/ModelTrait.php (hunk 3 above)
Comment (David-Denny) [reply src/Model/ModelTrait.php]: Fixed

File: test/ModelTraitTest.php [hunk 6]
```diff
@@ -49,23 +51,26 @@ public function testSetData($testData, $expectedProperties)
...
-                            $this->assertAttributeEquals($subValue, $subProp, $actualData[$i]);
+                            $assertGetter = "get" . ucfirst($subProp);
+                            $this->assertSame($subValue, $actualData[$i]->{$assertGetter}($subProp));
                         }
                     }
                 }
                 continue;
             }
-            $this->assertAttributeEquals($value, $prop, $modelTrait);
+            $getter = "get" . $prop;
```
Comment (downsider) [inline test/ModelTraitTest.php]: you'd need to run `ucfirst()` on `$prop` here, no?

File: src/Model/ModelTrait.php (hunk 4 above)
Comment (David-Denny) [reply src/Model/ModelTrait.php]: Fixed

File: src/Model/ModelTrait.php (hunk 5 above)
Comment (David-Denny) [reply src/Model/ModelTrait.php]: Fixed

File: test/ModelTraitTest.php [hunk 7]
```diff
@@ -49,23 +51,26 @@ public function testSetData($testData, $expectedProperties)
                 if (!empty($value["objectData"][0])) {
                     foreach ($value["objectData"] as $i => $objData) {
                         foreach ($objData as $subProp => $subValue) {
-                            $this->assertAttributeEquals($subValue, $subProp, $actualData[$i]);
+                            $assertGetter = "get" . ucfirst($subProp);
+                            $this->assertSame($subValue, $actualData[$i]->{$assertGetter}($subProp));
```
Comment (downsider) [inline test/ModelTraitTest.php]: The getter doesn't need an argument

Same on the change you made below

File: test/ModelTraitTest.php [hunk 8]
```diff
@@ -142,31 +148,23 @@ public function setDataProvider()
                     "camelCaseProp2" => "value2",
                 ]
             ],
-            [ // #3 set properties directly (no setter method)
-                [
-                    "noSetterProp" => "value1"
```
Comment (downsider) [inline test/ModelTraitTest.php:147]: Why did you remove this test?

File: test/ModelTraitTest.php (hunk 6 above)
Comment (David-Denny) [reply test/ModelTraitTest.php]: Yep, added. Looks like the tests still worked without it though

File: src/Model/ModelTrait.php [hunk 9]
```diff
@@ -151,10 +150,11 @@ private function getFirstParameter(string $method): \ReflectionParameter
...
     private function constructClasses(\ReflectionParameter $param, mixed $value, bool $isCollection = false): mixed
     {
-        $class = $param->getType() && !$param->getType()->isBuiltin()
-            ? new \ReflectionClass($param->getType()->getName()) : null;
+        $classType = $param->getType();
+        $class = $classType && is_a($classType, "ReflectionNamedType") && !$classType->isBuiltin()
+            ? new \ReflectionClass($classType->getName()) : null;
 
-        if (!empty($class)) {
+        if (is_a($class, "ReflectionClass")) {
```
Comment (downsider) [inline src/Model/ModelTrait.php]: I'd use `instanceof` instead of `is_a()`. You don't need the class name in a string

File: src/Model/ModelTrait.php (hunk 4 above)
Comment (downsider) [reply src/Model/ModelTrait.php]: Not quite. `isBuiltIn` is a method of `ReflectionNamedType` and you use that method before checking the class

Comment (downsider) [reply src/Model/ModelTrait.php]: Also, if you're checking the type (with `instanceof` (hint)) you don't need the boolean-ey check at the start

File: test/ModelTraitTest.php (hunk 7 above)
Comment (David-Denny) [reply test/ModelTraitTest.php]: Fixed

File: src/Model/ModelTrait.php (hunk 9 above)
Comment (David-Denny) [reply src/Model/ModelTrait.php]: Fixed

File: src/Model/ModelTrait.php (hunk 4 above)
Comment (David-Denny) [reply src/Model/ModelTrait.php]: Should be fixed now, switched to instanceof

File: test/ModelTraitTest.php (hunk 8 above)
Comment (David-Denny) [reply test/ModelTraitTest.php:147]: My mistake, restored the test

File: src/Model/ModelTrait.php [hunk 10]
```diff
@@ -60,19 +63,19 @@ protected function loadData(array $data, $update = false, $replaceCollections =
                     // if the setter's first parameter is an array, we need to check if this is a collection of objects
                     // in this case there should be an "addProperty" method on the class
                     $adder = "add$prop";
-                    if ($param->isArray() && method_exists($this, $adder)) {
+                    if ($param->getType() && $param->getType()->getName() === 'array' && method_exists($this, $adder)) {
```
Comment (downsider) [inline src/Model/ModelTrait.php]: Reckon we need the instanceof check here too. `getName()` only exists on `ReflectionNamedType`

File: src/Model/ModelTrait.php [hunk 11]
```diff
@@ -60,19 +63,19 @@ protected function loadData(array $data, $update = false, $replaceCollections =
...
 
-                    // if we have a collection and we're updating it ...
+                    // if we have a collection, and we're updating it ...
                     if (
                         $isCollection &&
                         $update &&
                         (
                             empty($replaceCollections) ||               // don't add if we're replacing all collections
-                            empty($replaceCollections[$this->mbLcfirst($prop)])  // don't add if this collection has been marked for replacing
+                            empty($replaceCollections[$this->mbLcfirst($prop)])  // don't add if this collection has been)
```
Comment (downsider) [inline src/Model/ModelTrait.php]: Think you might have accidentally removed half this comment :)

File: src/Model/ModelTrait.php (hunk 10 above)
Comment (David-Denny) [reply src/Model/ModelTrait.php]: As it's checking for an array, should I use is_array rather than instanceof?

File: src/Model/ModelTrait.php (hunk 11 above)
Comment (David-Denny) [reply src/Model/ModelTrait.php]: Restored

File: src/Model/ModelTrait.php (hunk 10 above)
Comment (downsider) [reply src/Model/ModelTrait.php]: The code is checking the typehint of a method argument via reflection, so `is_array` wouldn't apply here. `$param` should still be a `ReflectionNamedType` at this point

File: test/ModelTraitTest.php [hunk 12]
```diff
@@ -27,6 +28,7 @@ class ModelTraitTest extends \PHPUnit_Framework_TestCase {
      *
      * @param $testData
      * @param $expectedProperties
+     * @throws ModelException| \ReflectionException
```
Comment (downsider) [inline test/ModelTraitTest.php]: For consistency, each Exception should have it's own `@throws` declaration

Comment (David-Denny) [reply test/ModelTraitTest.php]: Separated the `@throws`

File: src/Model/ModelTrait.php (hunk 10 above)
Comment (David-Denny) [reply src/Model/ModelTrait.php]: Added instanceof check

File: test/Implementation/ModelTraitImplementation.php [hunk 13]
```diff
@@ -134,9 +134,9 @@ public function setCollectionProp(array $collectionProp)
     }
 
     /**
-     * @param ModelTraitImplementation $collectionProp
+     * @param ModelTraitImplementation|array $collectionProp
      */
-    public function addCollectionProp(ModelTraitImplementation $collectionProp)
+    public function addCollectionProp(ModelTraitImplementation|array $collectionProp)
```
Comment (downsider) [inline test/Implementation/ModelTraitImplementation.php]: No, this can't be an array. Why have you changed this?

File: src/Model/ModelTrait.php (hunk 10 above)
Comment (downsider) [reply src/Model/ModelTrait.php]: yeah, but not for the right class. Also `array` is not a class so you can't use `instanceof` on it anyway

remember you're not checking for an array, you're checking the object that `getType()` returns is the correct class

File: test/Implementation/ModelTraitImplementation.php (hunk 13 above)
Comment (David-Denny) [reply test/Implementation/ModelTraitImplementation.php]: Data set 5:
`[
                    "collectionProp" => [
                        $prop1Data,
                        $prop2Data,
                        $prop3Data
                    ]
]`
is an array when it's passed to `addCollectionProp` in the loadData function

Comment (David-Denny) [reply test/Implementation/ModelTraitImplementation.php]: In setCollectionProp, the elements of the input array are also arrays. I could convert them to instances of ModelTraitImplementations before calling addCollectionProp

Comment (downsider) [reply test/Implementation/ModelTraitImplementation.php]: it shouldn't be, it should loop over that array and add each entry individually

Comment (downsider) [reply test/Implementation/ModelTraitImplementation.php]: That conversion is meant to happen automatically, thats one of the main points of clay

Comment (downsider) [reply test/Implementation/ModelTraitImplementation.php]: I think this is failing because t's not detecting the class type due to the issue in the other open comment. 

You can definitely revert these "array" changes; the tests weren't in error, the code is.

Comment (David-Denny) [reply test/Implementation/ModelTraitImplementation.php]: Will do

File: src/Model/ModelTrait.php (hunk 10 above)
Comment (David-Denny) [reply src/Model/ModelTrait.php]: Should be correct now, the tests pass after I've removed the array changes

Comment (downsider) [reply src/Model/ModelTrait.php]: Not quite. You need to check the `getName()` returns `"array"`

Comment (David-Denny) [reply src/Model/ModelTrait.php]: Added the array check back in

File: src/Model/ModelTrait.php [hunk 14]
```diff
@@ -65,7 +65,11 @@ protected function loadData(array $data, bool $update = false, array $replaceCol
                     // in this case there should be an "addProperty" method on the class
                     $adder = "add$prop";
 
-                    if ($param->getType() instanceof \ReflectionNamedType && method_exists($this, $adder)) {
+                    $paramType = $param->getType();
+                    if ($paramType instanceof \ReflectionNamedType &&
```
Comment (downsider) [inline src/Model/ModelTrait.php]: If we're splitting to multi-line, this first condition should also be on it's own line

File: test/ModelTraitTest.php [hunk 15]
```diff
@@ -49,23 +54,31 @@ public function testSetData($testData, $expectedProperties)
                 if (!empty($value["objectData"][0])) {
                     foreach ($value["objectData"] as $i => $objData) {
                         foreach ($objData as $subProp => $subValue) {
-                            $this->assertAttributeEquals($subValue, $subProp, $actualData[$i]);
+                            $assertGetter = "get" . ucfirst($subProp);
+
+                            if (is_array($actualData[$i])) {
+                                $this->assertSame($subValue, $actualData[$i][$subProp]);
```
Comment (downsider) [inline test/ModelTraitTest.php]: This if statement should no longer be required

File: src/Model/ModelTrait.php (hunk 14 above)
Comment (David-Denny) [reply src/Model/ModelTrait.php]: Fixed

File: test/ModelTraitTest.php (hunk 15 above)
Comment (David-Denny) [reply test/ModelTraitTest.php]: Removed

File: composer.lock [hunk 16]
```diff
@@ -4,251 +4,1165 @@
...
-    "content-hash": "c9216234e474339c950da59f72180d0a",
+    "content-hash": "8402beff2b9f4fd299973a3774e5ead6",
     "packages": [
         {
             "name": "lexide/k-switch",
-            "version": "1.0.1",
+            "version": "2.0.0",
+            "source": {
+                "type": "git",
+                "url": "https://gitlab.ins.risk.regn.net/UKI/shared/php/packages/lexide/k-switch.git",
```
Comment (downsider) [inline composer.lock]: Ah, you need to disable the private packagist repo in your composer before running composer update; this gitlab URL is not public

File: src/Model/ModelTrait.php [hunk 17]
```diff
@@ -140,13 +152,17 @@ private function getFirstParameter($method)
...
+     * @throws \ReflectionException
      */
-    private function constructClasses(\ReflectionParameter $param, $value, $isCollection = false)
+    private function constructClasses(\ReflectionParameter $param, mixed $value, bool $isCollection = false): mixed
     {
-        $class = $param->getClass();
-        if (!empty($class)) {
+        $classType = $param->getType();
+        $class = ($classType instanceof \ReflectionNamedType) && !$classType->isBuiltin()
+            ? new \ReflectionClass($classType->getName()) : null;
```
Comment (downsider) [inline src/Model/ModelTrait.php]: `: null;` should be on it own line

File: composer.lock (hunk 16 above)
Comment (David-Denny) [reply composer.lock]: How do I do this?

File: src/Model/ModelTrait.php (hunk 17 above)
Comment (David-Denny) [reply src/Model/ModelTrait.php]: Fixed

File: composer.lock (hunk 16 above)
Comment (David-Denny) [reply composer.lock]: Nevermind, fixed
